File: App/Kairos-Calendar-Android-free/app/src/main/java/com/kairos_calendar/kairoscalendarfree/WebAppInterface.java
package com.kairos_calendar.kairoscalendarfree;

import android.os.Handler;
import android.util.Log;
import android.webkit.JavascriptInterface;
import android.widget.Toast;

public class WebAppInterface
{
    private final static String LOGTAG = "WebAppInterface";

    private MainActivity mainActivity;
    private DataManager dataManager;
    private Handler mHandler;

    WebAppInterface(MainActivity main, DataManager dataManager, Handler mHandler)
    {
        this.mainActivity = main;
        this.dataManager  = dataManager;
        this.mHandler     = mHandler;
    }

    @JavascriptInterface
    public void showToast(String toast)
    {
        Toast.makeText(mainActivity, toast, Toast.LENGTH_SHORT).show();
    }

    @JavascriptInterface
    public void openScanner()
    {
        Log.d(LOGTAG, "openScanner");

        Runnable startQrRunnable = new Runnable()
        {
            @Override
            public void run()
            {
                mainActivity.openScanner();
            }
        };

        mHandler.post(startQrRunnable);
    }

    @JavascriptInterface
    public void reload()
    {
        Log.d(LOGTAG, "reload");

        Runnable startQrRunnable = new Runnable()
        {
            @Override
            public void run()
            {
                mainActivity.reload();
            }
        };

        mHandler.post(startQrRunnable);
    }

    @JavascriptInterface
    public void openMaps(final String place)
    {
        Log.d(LOGTAG, "openMaps: " + place);

        Runnable startQrRunnable = new Runnable()
        {
            @Override
            public void run()
            {
                mainActivity.openMaps(place);
            }
        };

        mHandler.post(startQrRunnable);
    }

    @JavascriptInterface
    public String events()
    {
        return dataManager.events();
    }

    @JavascriptInterface
    public String add(String entry)
    {
        return dataManager.add(entry);
    }

    @JavascriptInterface
    public void delete(String uuid)
    {
        dataManager.delete(uuid);
    }

    @JavascriptInterface
    public void update(String uuid, String entry)
    {
        dataManager.update(uuid, entry);
    }

    @JavascriptInterface
    public String config()
    {
        return dataManager.config();
    }

    @JavascriptInterface
    public void putConfig(String json)
    {
        dataManager.putConfig(json);
    }

    @JavascriptInterface
    public void exeIntent()
    {
        Log.d(LOGTAG, "exeIntent");

        Runnable intent = new Runnable()
        {
            @Override
            public void run()
            {
                mainActivity.exeIntent();
            }
        };

        mHandler.post(intent);
    }

    @JavascriptInterface
    public void share(final String event)
    {
        Log.d(LOGTAG, "share: " + event);

        Runnable intent = new Runnable()
        {
            @Override
            public void run()
            {
                mainActivity.share(event);
            }
        };

        mHandler.post(intent);
    }

    @JavascriptInterface
    public void onBackPressed()
    {
        Log.d(LOGTAG, "onBackPressed");

        Runnable run = new Runnable()
        {
            @Override
            public void run()
            {
                mainActivity.backPressed();
            }
        };

        mHandler.post(run);
    }
}

File: App/Kairos-Calendar-IOS/Dragon-Calendar/WebAppInterface.swift
//
//  WebAppInterface.swift
//  Dragon-Calendar
//
//  Created by patrick zierahn on 21/10/16.
//  Copyright © 2016 patrick zierahn. All rights reserved.
//

import WebKit
import Foundation

class WebAppInterface: NSObject, WKScriptMessageHandler
{
    var mainView: ViewController
    
    init(mainView: ViewController)
    {
        self.mainView = mainView
    }
    
    static func getConfig(interface: WebAppInterface) -> WKWebViewConfiguration
    {
        let contentController = WKUserContentController()
        contentController.add(interface, name: "showToast")
        contentController.add(interface, name: "openScanner")
        contentController.add(interface, name: "putConfig")
        contentController.add(interface, name: "update")
        contentController.add(interface, name: "events")
        contentController.add(interface, name: "delete")
        contentController.add(interface, name: "config")
        contentController.add(interface, name: "reload")
        contentController.add(interface, name: "exeIntent")
        contentController.add(interface, name: "share")
        contentController.add(interface, name: "openMaps")

        let config = WKWebViewConfiguration()
        config.userContentController = contentController

        return config
    }
    
    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage)
    {
        if (message.name == "showToast")
        {
            NSLog("Javascript: " + (message.body as! String))
        }
        
        if (message.name == "openScanner")
        {
            NSLog("openScanner");
//            ViewController.openQrScanner(mainView: mainView);
            mainView.showQrView()
        }
        
        if (message.name == "putConfig")
        {
            DataManager.putConfig(config: (message.body as! String))
        }
        
        if (message.name == "update")
        {
            let entry = (message.body as! String)
            let id = Simple.getId(json: entry)
            
            DataManager.update(uuid: id, entry: entry)
        }
        
        if (message.name == "events")
        {
            DataManager.events()
        }
        
        if (message.name == "delete")
        {
            if let id = message.body as? String
            {
                DataManager.delete(uuid: id)
            }
        }
        
        if (message.name == "config")
        {
            let callback = (message.body as! String)
            DataManager.config(callback: callback)
        }
        
        if (message.name == "reload")
        {
            NSLog("reload")
            ViewController.updateWebView()
        }
        
        if (message.name == "exeIntent" && ViewController.intentString != "")
        {
            NSLog("exeIntent: " + ViewController.intentString)
            
            let call = "NativeInterface.intent('" + ViewController.intentString + "')"
            ViewController.myWebView.evaluateJavaScript(call, completionHandler: nil)
        }
        
        if (message.name == "share")
        {
            let event = (message.body as! String)
            mainView.share(event: event)
        }
        
        if (message.name == "openMaps")
        {
            let place = (message.body as! String)
            mainView.openMaps(place: place)
        }
    }
}
